proposals/distribution.md

```diff
@@ -107,16 +109,10 @@ The following entries should be added to the [scope table][scope]:
     Retrieving images covers the current “tag listing” (e.g. “what named manifests are in `library/busybox`?”), because tags are entries in the image format's [`manifests` array][manifests].
     Other tag-listing endpoints needed for backwards-compatibility are therefore in scope as well.
 
-    Repository actions and listing images within a repository are considered part of distribution policy or content management and are out of scope for this entry's per-image action.
-    For example, “what images are under `library/`?” is out of scope for this project.
-
     * What: Specifying a distribution method
     * In/Out/Future: In scope
     * Status: In progress (see opencontainers/distribution-spec)
     * Description: Define a protocol for image, manifest, config, and blob creation, retrieval, and deletion.
-        Listing repositories is a multi-repository action, which is out of scope for this entry.
-        Creating and deleting repositories are per-repository actions, which are out of scope for this entry.
-        Listing images within repositories is a per-repository action, which is out of scope for this entry.
```

Comment:
One could easily argue specifying a `/` in a path is a slippery slope.  

Listing an item is not the same as managing an item.  The whole point of a distribution spec is to enable content management. One can argue that enabling retrieval of a list of contents is content management, however I disagree. IMO retrieving a list of contents that have been pushed is enabling content management not providing content management.

Reply:
> One could easily argue specifying a / in a path is a slippery slope.

True, but backing that out would break compat with the established API.

> The whole point of a distribution spec is to enable content management. One can argue that enabling retrieval of a list of contents is content management, however I disagree. IMO retrieving a list of contents that have been pushed is enabling content management not providing content management.

To avoid getting lost in “enabling” vs. “providing”, I think we should focus on “what workflows should the coming spec enable?”.  That's a TOB-level scoping question.  We all agree it should enable push/pull of `wking/busybox` and constituent tags.  If you add additional language to the spec for other features (like the ones I was making out-of-scope), you'll want to make sure to label them as OPTIONAL for discovery servers that don't want to implement them.  Otherwise you burden those implementations in order to support the content-management API.

Reply:
Yes that's the key thing to specify here.. what workflows should the coming spec enable. Specifying what it won't enable is in and of itself a slippery slope :-) If there's a phrasing you want to add to ensure the resulting spec must delineate MUST items necessary to certain primary  workflows and further delineate as OPTIONAL the portions of the spec not necessary for the primary workflows that works for me.

Reply:
> Yes that's the key thing to specify here.. what workflows should the coming spec enable.

So what workflows do these repository endpoints enable?  Is the idea that folks can write clients for creating new repositories in an OCI registry before pushing images to it?  Or do you need repo create/delete for other purposes?

For listing images with a repository, you mentioned the “[I wonder what the rest of my team has created][1]” usecase.  If that's an intended usecase, then we should put that this proposal.  And the implementation would probably be along the lines of `/{repo}/actions`, since there are more types of activity than image creation/deletion (e.g. adding/removing tags from images).

I can't think of a usecase for “here are the names of all the images inside this repository” besides getting an overview of small repositories.  For browsing larger repositories, I'd expect an image count and a richer search.

>If there's a phrasing you want to add to ensure the resulting spec…

This scoping language is about whether the distribution project touches it at all.  I think things like REQUIRED vs. OPTIONAL are better discussed in the spec repo once it's created.

[1]: https://github.com/opencontainers/tob/pull/37#issuecomment-361736799

Reply:
> So what workflows do these repository endpoints enable? Is the idea that folks can write clients for creating new repositories in an OCI registry before pushing images to it? Or do you need repo create/delete for other purposes?

Good questions.. 

> For listing images with a repository, you mentioned the “I wonder what the rest of my team has created” usecase. If that's an intended usecase, then we should put that this proposal. And the implementation would probably be along the lines of /{repo}/actions, since there are more types of activity than image creation/deletion (e.g. adding/removing tags from images). 

Yes, I'm not sure what the value of distribution is if not to enable work flows with image sharing.

> I can't think of a usecase for “here are the names of all the images inside this repository” besides getting an overview of small repositories. For browsing larger repositories, I'd expect an image count and a richer search.

Yes, right after we agree a list is needed, we'll have the discussion about search and / or filtering the list. 

> This scoping language is about whether the distribution project touches it at all.

All is a big list.. I don't think we need to provide a hard list of everything that is not to be touched.

> I think things like REQUIRED vs. OPTIONAL are better discussed in the spec repo once it's created.

Agreed.

Reply:
> Yes, I'm not sure what the value of distribution is if not to enable work flows with image sharing.

The distribution is not going to be a complete solution.  Just like everyone seems to agree that discovery is out of scope here, I think this monitoring usecase is best handled separately.

> Yes, right after we agree a list is needed, we'll have the discussion about search and / or filtering the list.

Search engines often decouple nicely from covered content.  Do you see much cross-cutting here?

> I don't think we need to provide a hard list of everything that is not to be touched.

I'm trying to stick to categories folks have mentioned since #34 or which are [already mentioned in the Docker spec][1], since those seem like the most likely scope-creep directions.

[1]: https://github.com/docker/distribution/blob/6fca8d6e6713acbdf3f9ca40cf6370fc5ee5ee53/docs/spec/api.md#listing-repositories

Reply:
> Search engines often decouple nicely from covered content. 

Define decouple, the search engines I'm aware of all have a way to iterate over the covered content.

> I'm trying to stick to categories folks have mentioned since #34 or which are already mentioned in the Docker spec, since those seem like the most likely scope-creep directions.

> The distribution is not going to be a complete solution.

The docker spec you cite above looks to be a complete solution. I get that this specification may be a subset of that specification but that does not mean it must be an incomplete solution.

Reply:
> Define decouple, the search engines I'm aware of all have a way to iterate over the covered content.

Google provides a way to iterate over pages of matches, but the Google search API is not defined in an HTTP or HTML RFC.  They walk that content and then provide an index to it.  Similarly, you could have a registry-search engine that walked an engine using some boring internal mechanism and then provided a standardized (or not) search API.  But we don't need to define either the search API or the registry-walking API in order to support standardized image push/pull.  I'd rather bite those components off one at a time (or leave them unstandardized) instead of rolling them all together into a single spec.

The only reason I could see for making registry search part of the image push/pull spec would be if there were strong synchronicity requirements (e.g. all searches after a push MUST include that image in the result set and all searches after a delete MUST NOT include that image).  But search is for browsing people, not machines, so I don't think we need that sort of synchronicity.

> > The distribution is not going to be a complete solution.
>
> The docker spec you cite above looks to be a complete solution…

Complete for what task?  Even if your goal is just:

```console
$ my-tool pull wking/busybox
```

something is going to have to tell your tool “hey, you might want to look in https://index.docker.io/v2”.  That leg is discovery (e.g. [oci-discovery][1] and [parcel][2]), and folks seem to agree that it's [out of scope for this project][3].  By also excluding registry operations, I'm just trying to focus this project a little more tightly on image push/pull.  Yes that leaves out search; but I don't think leaving search our of *this* spec is a big problem.

[1]: https://github.com/xiekeyang/oci-discovery
[2]: https://github.com/cyphar/parcel
[3]: https://github.com/opencontainers/tob/issues/34#issuecomment-350529321

Reply:
> By also excluding registry operations, I'm just trying to focus this project a little more tightly on image push/pull. 

Ok but this is the section that talks about the registry operations that are in scope not the ones that are out of scope. 

What do you think of splitting `Specifying a distribution method` into multiple entries, one section for content distribution operations and another section for the rest of the registry operations that are necessary for implementing a current state of the art OCI container image registry api?

Reply:
> Ok but this is the section that talks about the registry operations that are in scope not the ones that are out of scope.

This section is about making “distribution” in-scope and defining distribution as operations that take:

* A registry implementing the coming distribution API.
* An image or more granular object within (or about to be within) that registry.

I'm trying to place operations that take higher-level objects (e.g. “repositories” as collections of images) outside of distribution.  They're discovery, or image policy, or search, or signing/naming, etc.  I don't see the point of collecting a handful of those together and calling it a registry.

> What do you think of splitting “Specifying a distribution method” into multiple entries, one section for content distribution operations…

I think this is what I'm aiming at [already][1].

> … and another section for the rest of the registry operations that are necessary for implementing a current state of the art OCI container image registry api?

What are these?  Just search?  I think the [catalog endpoint][2] is the only current endpoint I'd consider out-of-scope, and the only hit for that in Moby I see for it is the vendored registry API:

```console
moby$ git grep _catalog
vendor/github.com/docker/distribution/registry/api/v2/descriptors.go:           Path:        "/v2/_catalog",
```

The catalog endpoint landed in docker/distribution#653, where the use-case was seeding an external index which is then kept up to date via an events API (maybe [this][3]?).  But from the issue it seems more like a backend hack than part of the xternal user's API.

[1]: https://github.com/opencontainers/tob/blame/5dcd80a038fd2214524cea0ffbf9cfaa9390900b/proposals/distribution.md#L104
[2]: https://github.com/docker/distribution/blob/6fca8d6e6713acbdf3f9ca40cf6370fc5ee5ee53/docs/spec/api.md#listing-repositories
[3]: https://github.com/docker/distribution/blob/6fca8d6e6713acbdf3f9ca40cf6370fc5ee5ee53/docs/configuration.md#notifications

Reply:
Heh, and from the docker/distribution#653 [discussion][1]:

> If this is not meant for end-user consumption, then maybe it should go in a separate doc covering administrative endpoints?

which is pretty much exactly my position, except in a 2015-07-13 comment before the feature landed in docker/distribution.  I don't see any follow up on that comment there, but perhaps we can address it now before we drag it along into the OCI spec.


[1]: https://github.com/docker/distribution/pull/653#issuecomment-121099454

Reply:
> "and another section for the rest of the registry operations that are necessary for implementing a current state of the art OCI container image registry api?" What are these?

From the discussion we had in Austin the idea was to accept the docker registry v2 api, as the distribution api or fairly close to that. I'm having a hard time mapping docker's api to the "out of scope vs in scope" discussion.  Would be nice if there was a "details.md" that highlighted what is not in scope from the v2 detail table.

Reply:
From the [the detail table][1], the endpoints all take the image `<name>` (and so are explicitly in-scope) except for:

*  `/v2/`, but I'm fine with version-detection.
* `/v2/_catalog`, which I thing is out of scope.  I think it falls pretty clearly under the "Listing repositories..." restriction you're dropping here, but if a local table would help clarify, that's fine with me.

https://github.com/docker/distribution/blob/6fca8d6e6713acbdf3f9ca40cf6370fc5ee5ee53/docs/spec/api.md#detail

Reply:
> Haven't seen a definitive is discussion on is `<name>` in or out of scope, only that it's out of scope with regard to repository (catalog) actions.

Yeah, we can probably clarify that more.  The Docker spec [has][1]:

> All endpoints will be prefixed by the API version and the repository name:
>
>     /v2/<name>/

and then [later on][2]:

> An "image" is a combination of a JSON manifest and individual layer files. The
process of pulling an image centers around retrieving these two components.
>
> The first step in pulling an image is to retrieve the manifest. For reference,
the relevant manifest fields for the registry are the following:
>
>  field    | description                                    |
> ----------|------------------------------------------------|
> name      | The name of the image.                         |
> tag | The tag for this version of the image. |

That isn't particularly clear on whether `<name>` is a repository (as the first excerpt suggests) or an image (as the second excerpt suggests).  I've been using “image” to mean “a collection of tagged/named references”.  The image-spec phrasing for that is an “[image index][3]”.  I think operations that take an image-index name as an argument are in-scope, because image-spec's index object can represent them in `manifests`.  Operations that take or return collections of image-indexes (which I've been referring to as “repositories”) are out-of-scope.

> Also haven't seen a definitive is `GET /v2/<name>/tags/list` in or out…

I'd been arguing to put that in-scope based on the distinction between “image” (-index) and “manifest” [here][4].  I also discuss tag-listing [here][5], although that's currently outside of the scope-table entry.

[1]: https://github.com/docker/distribution/blame/6fca8d6e6713acbdf3f9ca40cf6370fc5ee5ee53/docs/spec/api.md#L223
[2]: https://github.com/docker/distribution/blame/6fca8d6e6713acbdf3f9ca40cf6370fc5ee5ee53/docs/spec/api.md#L391
[3]: https://github.com/opencontainers/image-spec/blob/v1.0.1/image-index.md
[4]: https://github.com/opencontainers/tob/blame/5dcd80a038fd2214524cea0ffbf9cfaa9390900b/proposals/distribution.md#L116
[5]: https://github.com/opencontainers/tob/blame/5dcd80a038fd2214524cea0ffbf9cfaa9390900b/proposals/distribution.md#L107-L108

Reply:
Sounds like we are in full agreement on these details then. Just a matter of terminology and approach. That we can be in agreement on the detail but not the proposal is why I think less is more in the proposal wrt to what should be drawn as out of scope to reduce confusion. Well unless we are going to go the extra mile and describe it to the level of detail in the proposal, which then arguable makes it tldr.

Reply:
> That we can be in agreement on the detail but not the proposal is why I think less is more in the proposal wrt to what should be drawn as out of scope (to reduce confusion).

Ye of little faith ;).  I've filed #46 with an attempt at less ambiguous wording.

Reply:
well done @wking I like new version.  Tower of babel has been brought down.. for the moment anyway :-)